urlfd018a75f7_highlanderkev_StockExVirtuoso_tgz-pJ8-driver_Phase4J8/src/publishers/TickerPublisher.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * TickerPublisher.java
+ */
+package publishers;
+
+import client.User;
+import exceptions.ExceptionHandler;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import price.Price;
+import price.PriceFactory;
+
+/**
+ * This class implements the "singleton" design pattern, and only one should
+ * ever be created.<br> This class handles sending out tickers to users about
+ * the market.
+ *
+ * @author Kevin Patrick Westropp
+ * @version StockExVirtuoso.Version.1.0
+ */
+public class TickerPublisher implements Publisher {
+
+    /**
+     * Delegate - will refer to myPublisher implementation object
+     */
+    private static PublisherImpl myPublisher;
+    /**
+     * Single instance of the MessagePublisher
+     */
+    private static TickerPublisher instance;
+    /**
+     * HashMap holds most up to date stock values
+     */
+    private static HashMap<String, Price> stocks = new HashMap();
+
+    /**
+     * Private Constructor for single instance construction.
+     */
+    private TickerPublisher() {
+    }
+
+    /**
+     * Public synchronized get Instance method which returns the instance of the
+     * TickerPublisher.
+     *
+     * @return TickerPublisher instance
+     */
+    public synchronized static TickerPublisher getInstance() {
+        if (instance == null) {
+            instance = new TickerPublisher();
+            myPublisher = new PublisherImpl();
+        }
+        return instance;
+    }
+
+    /**
+     * Public synchronized publish ticker method which sends out tickers about
+     * the market.
+     *
+     * @param product symbol
+     * @param p price object
+     * @throws Exception
+     */
+    public synchronized void publishTicker(String product, Price p) throws Exception {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.